### PR TITLE
Fix dynamic array spec

### PR DESCRIPTION
**Pull request by supermassive**
Fixes dynamic array spec. 
Elements should be encoded as tuple so everything after `enc(k)` should be same as for fixed size array above

### CODE REVIEW

File: docs/abi-spec.rst
Comment (ekpyron) [inline docs/abi-spec.rst]: Without looking at it and thinking through it in all detail, I actually think the previous version is correct and the change *makes* this incorrect.
``[X[0], ..., X[k-1]]`` is meant to be a fixed size array, which will then recurse back to the previous clause, ultimately resulting in ``(X[0], ..., X[k-1])``.

A single-element tuple of a statically-sized array, i.e. the version after the change, can actually be encoded differently (e.g. in case the base type is dynamic).

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: docs/abi-spec.rst
Comment (ekpyron) [reply docs/abi-spec.rst]: It would be an option to change this to ``enc((X[0], ..., X[k-1]))`` (note the missing ``[ ]`` compared to the version in the PR).
But if we did that, we'd also need to change the comment below.

Here one could argue either way - to me it seems nicer to reduce dynamic arrays to static arrays first, and only then to tuples, rather than directly to tuples - but we don't actually formally introduce the ``[...]`` notation for static arrays, do we? If we don't formally introduce that notation, directly moving to a tuple may indeed be simpler.

File: docs/abi-spec.rst
Comment (ekpyron) [inline docs/abi-spec.rst]: If we change the formula to a tuple, I'd suggest something like

```suggestion
  i.e. it is encoded as if it were a tuple with ``k`` elements of the same type (resp. an array of static size ``k``), prefixed with
  the number of elements.
```

here

File: docs/abi-spec.rst
Comment (supermassive) [reply docs/abi-spec.rst]: My bad. I meant it to match `enc((X[0], ..., X[k-1]))` above but somehow failed to remove `[]` brackets.

Comment (supermassive) [reply docs/abi-spec.rst]: @ekpyron PTAL

**Review by ekpyron — APPROVED** [review]
Looks good to me now, thank you very much!

**Comment by supermassive** [comment]
@ekpyron what is the process of making PRs merged to dev branch? I have no write access.

**Comment by ekpyron** [comment]
I was mainly waiting for CI to pass and whether anyone else wants to weigh in on this - but CI has passed and the change is good, so I'll merge now :-).

**Pull request merged**
